spanner/google/cloud/spanner/streamed.py

```diff
@@ -122,17 +130,50 @@ def _merge_values(self, values):
             field = self.fields[index]
             self._current_row.append(_parse_value_pb(value, field.type))
             if len(self._current_row) == width:
-                self._rows.append(self._current_row)
+                self._pending_rows.append(self._current_row)
                 self._current_row = []
 
+    def _flush_pending_rows(self):
+        """Helper for :meth:`consume_next`."""
+        flushed = self._pending_rows[:]
+        self._pending_rows[:] = ()
+        self._complete_rows.extend(flushed)
+
+    def _do_restart(self):
+        """Helper for :meth:`consume_next`."""
+        self._pending_chunk = None
+        self._pending_rows[:] = ()
+        self._current_row[:] = ()
+
+        if self._resume_token:
+            self._response_iterator = self._restart(self._resume_token)
+        else:
+            self._response_iterator = self._restart()
```

Comment:
What if you have already yielded some rows to the user but there is no resume token. In that case you can't just do a restart from beginning. You need to differentiate between 3 cases here:
1. There is a resume token.
2. There is no resume token but you have not yielded anything to the user.
3. There is no resume token but you have yielded one or more rows to the user.

In 1 and 2 it is safe to resume. In 3 you need to just raise an error to the user.

Reply:
We no longer yield rows to the user until we receive either a token or the last partial result set (the one with stats).  The not-yet-yielded rows are in the `self._pending_rows` attribute, cleared in line 145.  Those rows get copied into `self._completed_rows` via `self._flush_pending_rows()`, which is invoked only in those two cases (and the oddball edge case that the streaming iterator raises `StopIteration` without sending the last PRS).

Reply:
But even after you have yielded rows to the user, you might get more PRS's without resume tokens.
For eg say for a query with single string column, you get
PRS1: value: "hello", chunked: false, resume_token:"blah".
PRS2: value: "world", chunked: false, resume_token:Null

You will yield row2 to user after seeing PRS1. But PRS2 does not have resume token and if stream breaks after that you will again resume from beginning, thus yielding row1 to the user again.

Reply:
:) How do you strikethrough text here?

Reply:
`~` is the strikethrough character in Markdown.

Reply:
> How do you strikethrough text here?

Use "~~" on either side.